src/libraries/icubmod/bcbBattery/bcbBattery.cpp

```diff
@@ -60,174 +60,95 @@ bool BcbBattery::open(yarp::os::Searchable& config)
+    batteryReader->screenEnable = group_general.check("screen", Value(0), "enable/disable the screen output").asBool();
+    batteryReader->silenceSyncWarnings = group_general.check("silence_sync_warnings", Value(0), "enable/disable the print of warnings in case of sync errors.").asBool();
 
-    this->isClosing = false;
-    PeriodicThread::start();
+    //start the thread
+    batteryReader->start();
     return true;
 }
 
 bool BcbBattery::close()
 {
-    isClosing = true;
+    //stop the thread
+    batteryReader->stop();
 
-    {
-        lock_guard<mutex> lck(mtx);
-
-        //stop the thread
-        PeriodicThread::stop();
-
-        char c = 0x00;
-        if (pSerial)
-        {
-            bool ret = pSerial->send(&c, 1);
-            if (ret == false) { yError("BcbBattery problems while stopping the transmission");}
-        }
-
-        //stop the driver
-        driver.close();
-    }
+    //stop the driver
+    driver.close();
 
     return true;
 }
 
-bool BcbBattery::threadInit()
+bool batteryReaderThread::threadInit()
 {
-    battery_info = "icub battery system v1.0";
-    battery_voltage     = 0.0;
-    battery_current     = 0.0;
-    battery_charge      = 0.0;
     timeStamp = yarp::os::Time::now();
 
-    //start the transmission
-    char c = 0x01;
     if (pSerial)
     {
-        bool ret = pSerial->send(&c, 1);
-        if (ret == false) { yError("BcbBattery problems starting the transmission"); return false; }
-
-        //empty the buffer
-        char c = 0;
-        int r = 0;
-        do
-        {
-            r = pSerial->receiveChar(c);
-        } while (r != 0);
+        yInfo("BcbBattery starting transmission");
+        startTransmission();
+        yInfo("BcbBattery started successfully");
     }
     else
     {
         yError("BcbBattery pSerial == NULL");
         return false;
     }
 
-
     return true;
 }
 
-void BcbBattery::run()
+void batteryReaderThread::run()
 {
     double timeNow=yarp::os::Time::now();
-    lock_guard<mutex> lck(mtx);
-
-    //if 100ms have passed since the last received message
-    if (timeStamp+0.1<timeNow)
-    {
-        //status=IBattery::BATTERY_TIMEOUT;
-    }
 
     //read battery data.
-    serial_buff[0] = 0;
-
-    int output[10] = {0};
-    bool syncError = false;
-
     if (pSerial)
     {
-        size_t i = 0;
-        bool sync0 = false;
-        bool syncr = false;
-        bool syncn = false;
-        do
+    search_h:
+        output[8] = pSerial->receiveChar(serial_buff[8]);
+        if (output[8] == 0 || serial_buff[8] != '\r')
         {
-            output[0] = pSerial->receiveChar(serial_buff[0]);
-            sync0 = output[0] && (serial_buff[0] == '\0');
-            ++i;
-        } while (!(sync0) && (i < 20) && !isClosing);
-
-        if (isClosing)
-        {
-            return;
+            if (!silenceSyncWarnings) yWarning("BcbBattery sync error r");
+            if (closeRequested) return;
+            goto search_h;
```

Comment:
The problem with the ``goto`` with respect to the previous implementation is that in case of problems, the buffer is never printed, neither in verbose mode. For example, if, for some reason, the module does not receive any ``\r`` (and we know that this is possible if the connection is not configured correctly), we have no chance to debug. It just keep printing the error at high rate (CPU rate). This is also very undesirable if the module starts in the main ``yarprobotinterface``.

With the previous logic, after 20 attempts it gives up and finishes the loop, printing only one warning per cycle.